Fix error in JFile::getExt

mavik opened the pull request:
JFile::getExt works incorrect if file does not have extension.

mavik committed:
Fix error in JFile::getExt

JFile::getExt works incorrect if file does not have extension.

Fedik commented:
Please add description how to reproduce the issue, and how to test the fix.

marenkov commented:
You can not reproduce this problem in the CMS, only in an  extension. But you can write php-code that uses it: $ext = JFile::getExt ('test');
Result: $ext = "est'
Right result: $ext = ''

Minei3oat commented:
Maybe a `false` is a better return value, if there is no extension. ` ` is returned if there is a dot as last letter in the string. So `false` would indicate that there is no dot.

Minei3oat commented:
I've tested `JFile::getExt('test.')`. It returns `false`, but it must be ` `.

Minei3oat commented on libraries/joomla/filesystem/file.php:
```diff
@@ -27,9 +27,9 @@ class JFile
-
-		return substr($file, $dot);
+		$dot = strrpos($file, '.');
+		if ($dot === false) { return ''; }
```
There is a code style problem ... https://joomla.github.io/coding-standards/?coding-standards/chapters/php.md

Minei3oat replied:
There is a code style problem ... https://joomla.github.io/coding-standards/?coding-standards/chapters/php.md

This should return `false` because there is no extension.

Minei3oat commented on libraries/joomla/filesystem/file.php:
```diff
@@ -27,9 +27,9 @@ class JFile
-		return substr($file, $dot);
+		$dot = strrpos($file, '.');
+		if ($dot === false) { return ''; }
+		return substr($file, $dot + 1);
```
This returns `false`  prior PHP7, but it should be ` ` because there is an extension.

ggppdk commented:
I confirm the issue and the fix, 
@marenkov , Joomla code must follow specific code styling

```if ($dot === false)
{
	return ''; 
}```


About
> Maybe a false is a better return value, if there is no extension. is returned if there is a dot as last letter in the string. So false would indicate that there is no dot.

Filename with trailing dot and without extension is usually not valid,
although some programs might create it, e.g. if you create such a name in windows it will be difficult to use or delete it ...

zero-24 commented:
What about

```
 	/**
 	 * Gets the extension of a file name
 	 *
 	 * @param   string  $file  The file name
 	 *
 	 * @return  string  The file extension
 	 *
 	 * @since   11.1
  	 */
  	public static function getExt($file)
  	{
  	  	return strtolower(pathinfo($file, PATHINFO_EXTENSION));
  	}
```

Based on: https://github.com/joomla-framework/filesystem/blob/master/src/Stream.php#L243

mavik committed:
Fix type of results

Return False if there is not dot, and '' if dot is last symbol.
I do not very like (string) substr, but it is very simple code and it works correct.

Minei3oat commented:
Nearly perfect, but there is the code style issue...
The following line condenses your two lines to one and solves the code style issue:
```php
return $dot === false ? false : (string) substr($file, $dot + 1);
```

ggppdk commented:
Recently i was trying to remember why i was not using
```
JFile::getExt
```

and instead i am using:
pathinfo($file, PATHINFO_EXTENSION)

@zero-24 
just, we cannot use strlower(), because 
- it will be change of behaviour for the method
- it is also incorrect to force it, the caller should get the true extension without any lowercasing, the caller will do it if needed, (current code does it already when needed)

about 
```
return $dot === false ? false : (string) substr($file, $dot + 1);
```

it seems good, but better not rediscover the wheel, for no good reason,
instead just use:
```
return pathinfo($file, PATHINFO_EXTENSION);
```

zero-24 commented:
Thanks @ggppdk good spot. I Agree with `return pathinfo($file, PATHINFO_EXTENSION);` instead of the current code ;)

mavik committed:
Update file.php

marenkov commented:
`return pathinfo($file, PATHINFO_EXTENSION);` looks good, but it returns string(0) for 'test.' and 'test' (instead false). 
If it is not problem, it is the best solution.

Minei3oat commented:
@ggppdk please use \`\`\`php instead of ```, because with the php the code is highlighted, wich makes it easier to read :)

ggppdk commented:
> @ggppdk please use ```php instead of ```, because with the php the code is highlighted, wich makes it easier to read :)

Done, thanks, i guess, i need to recheck gihub documentation

Minei3oat commented on libraries/joomla/filesystem/file.php:
```diff
@@ -28,7 +28,9 @@ class JFile
 	{
 		$dot = strrpos($file, '.');
-		if ($dot === false) { return false; }
+		if ($dot === false) {
```
The `{` must be in a new line ;)
or use the shorter version mentioned above.

marenkov replied:
It's crazy style. Why Joomla doesn't use PSR-2? :( 
But OK.
Can I use `@return string | boolean` in comments for function?

mavik committed:
Update file.php

Minei3oat commented:
I have tested this item :white_check_mark: successfully on 7c236a4a686b4adca2f74b069dcc7a36d65c32c0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12166">issues.joomla.org/joomla-cms/12166</a>.</sub>

zero-24 committed:
Add missing clean lines for PHPCS

zero-24 reviewed (approved):
Ok after the last CS commit.

ggppdk reviewed (approved):
Tested works as described, now, when no extension it returns false

ggppdk commented:
I have tested this item :white_check_mark: successfully on 7eab124596660278708d36c0d40b67efda0da59c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12166">issues.joomla.org/joomla-cms/12166</a>.</sub>

zero-24 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12166">issues.joomla.org/joomla-cms/12166</a>.</sub>

rdeutz commented:
Sorry guys, but I don't think we should do it this way. I don't like the different return types

```
JFile::getExt('whatever')
``` 

should return an empty string 

```
JFile::hasExt('whatever')
``` 

should return true or false (we don't have such a function atm and I am not saying we should make one :-))

rdeutz commented:
set to needs review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12166">issues.joomla.org/joomla-cms/12166</a>.</sub>

ggppdk commented:
@rdeutz 
> don't like the different return types
> should return an empty string 

Ok then, lets not make 2 changes in this PR, (1 fix + 1 behaviour change)

Let it return string as before, and ... only fix the original reported bug:
- Lets make it return empty string when no extension exists (= the original reported bug)

```php
// Check for no extension
if ($dot === false)
{
	return '';
}
```

marenkov commented:
In anyway current realization is wrong. We have two version of fix:
My version - we need select what is returned for files without extension and with dot at the end.
pathinfo - always returns string(0) if extension is empty. Not bad.
Somebody should select and fix it.

rdeutz commented:
I made a decision as maintainer, return an empty string instead of false

mavik committed:
Update file.php

marenkov commented:
OK. I've changed false to empty string.

rdeutz commented:
Thanks back to RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12166">issues.joomla.org/joomla-cms/12166</a>.</sub>

zero-24 commented:
:+1:
